src/librustc_typeck/coherence/overlap.rs

```diff
@@ -38,16 +38,23 @@ impl<'cx, 'tcx> OverlapChecker<'cx, 'tcx> {
         // check_for_overlapping_impls_of_trait() check, since that
         // check can populate this table further with impls from other
         // crates.
-        let trait_def_ids: Vec<ast::DefId> =
-            self.tcx.trait_impls.borrow().keys().map(|&d| d).collect();
-
-        for trait_def_id in trait_def_ids.iter() {
-            self.check_for_overlapping_impls_of_trait(*trait_def_id);
+        let trait_def_ids: Vec<(ast::DefId, Vec<ast::DefId>)> =
+            self.tcx.trait_impls.borrow().iter().map(|(&k, v)| {
+                let mut impls = v.borrow().clone();
+                if let Some(neg_impls) = self.tcx.trait_negative_impls.borrow().get(&k) {
+                    impls.push_all(neg_impls.borrow().as_slice());
+                }
+                (k, impls)
```

Comment:
(by merging `trait_impls` and `trait_negative_impls` this won't be needed anymore)